{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "PROJECT_DIR = '/home/shuang/projects/development_eqtm'\n",
    "tss_filepath = '/home/shuang/projects/development_eqtm/data/features/TSSDistance/Homo_sapiens.GRCh37.71.gtf'\n",
    "eqtm_filepath = '/home/shuang/projects/development_eqtm/data/eqtmZscores/withExpressionTSSMethy/2017-12-09-eQTLsFDR-et0_withExpressionTssMethy.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_TSS_basedOnProbeName_toEQTMFile(eqtm_filepath,tss_raw,\n",
    "                                      eqtm_savepath=None):\n",
    "    '''\n",
    "    add TssDistance to eqtm file\n",
    "    INPUT:\n",
    "        eQTM_path, string, path to eqtmZscore file\n",
    "        tss_filepath, string, path to TssSite file\n",
    "        save_path, string, path to save the new eqtmZscore_withTSSDistance file\n",
    "    OUTPUT:\n",
    "        eQTMs, pandas dataframe\n",
    "    '''\n",
    "    # tss_raw = read_tss_data(tss_filepath)\n",
    "    # reading the eQTMs\n",
    "    eQTMs = pd.read_csv(eqtm_filepath,sep=',')\n",
    "\n",
    "    # extract gene name from geneInfo for tss file\n",
    "    def findGeneName(item):\n",
    "        item = [thing for thing in list(filter(None,item.strip().split(\";\")))][0]\n",
    "        name = item.replace('\"','').replace(';','').strip().split(' ')[1]\n",
    "        return name\n",
    "    tss_raw['geneName'] = tss_raw['geneInfo'].apply(findGeneName)\n",
    "\n",
    "    # find the tss sites for each gene in the tss file\n",
    "    groupbyTss = tss_raw.groupby('geneName').agg({\n",
    "        'chr':lambda x: x.unique(),\n",
    "        'startSite':np.min,\n",
    "        'endSite':np.max,\n",
    "        'strand':lambda x: x.unique()\n",
    "    })\n",
    "    def findTssSite(series):\n",
    "        if series[3] == '-':\n",
    "            return series[2]\n",
    "        else:\n",
    "            return series[1]\n",
    "    groupbyTss['TssSite'] = groupbyTss.apply(findTssSite,axis=1)\n",
    "\n",
    "    # add tss sites and tss distance to the eqtm file\n",
    "    def mapSite(row):\n",
    "        return groupbyTss.loc[row]['TssSite']\n",
    "    def calculateDis(row):\n",
    "        return abs(row[0]-row[1])\n",
    "    def findChr(row):\n",
    "        return groupbyTss.loc[row]['chr']\n",
    "    def checkChr(row):\n",
    "        if str(row[0])==str(row[1]):\n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    eQTMs['TssSite'] = eQTMs['ProbeName'].apply(mapSite)\n",
    "    eQTMs['chr'] = eQTMs['ProbeName'].apply(findChr)\n",
    "    eQTMs['TssDistance'] = eQTMs[['SNPChrPos','TssSite']].apply(calculateDis,axis=1)\n",
    "    eQTMs['checkChr'] = eQTMs[['chr','SNPChr']].apply(checkChr,axis=1)\n",
    "    # check whether they are from the same chromosome\n",
    "    assert len(eQTMs['checkChr'].unique()) == 1\n",
    "\n",
    "    if eqtm_savepath:\n",
    "        # save the eQTM file\n",
    "        eQTMs.to_csv(eqtm_savepath,index=False)\n",
    "        print('Saved eQTM file to: ',eqtm_savepath)\n",
    "\n",
    "    return eQTMs\n",
    "\n",
    "def read_tss_data(tss_filepath):\n",
    "    '''\n",
    "    read tss file from tss_filepath\n",
    "    '''\n",
    "    colnames = ['chr','regionFunction','regionType','startSite',\n",
    "                'endSite','score','strand','sthunknown','geneInfo']\n",
    "    dtype = {'chr':object,'regionFunction':object,'regionType':object,\n",
    "             'startSite':int,'endSite':int,'score':object,\n",
    "             'strand':object,'sthunknown':object,'geneInfo':object}\n",
    "    tss_raw = pd.read_csv(tss_filepath,sep='\\t',header=None,\n",
    "                          names=colnames,dtype=dtype)\n",
    "    return tss_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "    tss_raw = read_tss_data(tss_filepath) # slow..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "    eQTMs = pd.read_csv(eqtm_filepath,sep=',')\n",
    "\n",
    "    # extract gene name from geneInfo for tss file\n",
    "    def findGeneName(item):\n",
    "        item = [thing for thing in list(filter(None,item.strip().split(\";\")))][0]\n",
    "        name = item.replace('\"','').replace(';','').strip().split(' ')[1]\n",
    "        return name\n",
    "    \n",
    "    tss_raw['geneName'] = tss_raw['geneInfo'].apply(findGeneName)\n",
    "\n",
    "    # find the tss sites for each gene in the tss file\n",
    "    groupbyTss = tss_raw.groupby('geneName').agg({\n",
    "        'chr':lambda x: x.unique(),\n",
    "        'startSite':np.min,\n",
    "        'endSite':np.max,\n",
    "        'strand':lambda x: x.unique()\n",
    "    })\n",
    "    def findTssSite(series):\n",
    "        if series[3] == '-':\n",
    "            return series[2]\n",
    "        else:\n",
    "            return series[1]\n",
    "    groupbyTss['TssSite'] = groupbyTss.apply(findTssSite,axis=1)\n",
    "\n",
    "#     # add tss sites and tss distance to the eqtm file\n",
    "#     def mapSite(row):\n",
    "#         return groupbyTss.loc[row]['TssSite']\n",
    "#     def calculateDis(row):\n",
    "#         return abs(row[0]-row[1])\n",
    "#     def findChr(row):\n",
    "#         return groupbyTss.loc[row]['chr']\n",
    "#     def checkChr(row):\n",
    "#         if str(row[0])==str(row[1]):\n",
    "#             return True\n",
    "#         else:\n",
    "#             return False\n",
    "#     eQTMs['TssSite'] = eQTMs['ProbeName'].apply(mapSite)\n",
    "#     eQTMs['chr'] = eQTMs['ProbeName'].apply(findChr)\n",
    "#     eQTMs['TssDistance'] = eQTMs[['SNPChrPos','TssSite']].apply(calculateDis,axis=1)\n",
    "#     eQTMs['checkChr'] = eQTMs[['chr','SNPChr']].apply(checkChr,axis=1)\n",
    "#     # check whether they are from the same chromosome\n",
    "#     assert len(eQTMs['checkChr'].unique()) == 1\n",
    "\n",
    "#     if eqtm_savepath:\n",
    "#         # save the eQTM file\n",
    "#         eQTMs.to_csv(eqtm_savepath,index=False)\n",
    "#         print('Saved eQTM file to: ',eqtm_savepath)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>startSite</th>\n",
       "      <th>endSite</th>\n",
       "      <th>strand</th>\n",
       "      <th>TssSite</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneName</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000000003</th>\n",
       "      <td>X</td>\n",
       "      <td>99883667</td>\n",
       "      <td>99894988</td>\n",
       "      <td>-</td>\n",
       "      <td>99894988</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000005</th>\n",
       "      <td>X</td>\n",
       "      <td>99839799</td>\n",
       "      <td>99854882</td>\n",
       "      <td>+</td>\n",
       "      <td>99839799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000419</th>\n",
       "      <td>20</td>\n",
       "      <td>49551404</td>\n",
       "      <td>49575092</td>\n",
       "      <td>-</td>\n",
       "      <td>49575092</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000457</th>\n",
       "      <td>1</td>\n",
       "      <td>169821804</td>\n",
       "      <td>169863408</td>\n",
       "      <td>-</td>\n",
       "      <td>169863408</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000000460</th>\n",
       "      <td>1</td>\n",
       "      <td>169631245</td>\n",
       "      <td>169823221</td>\n",
       "      <td>+</td>\n",
       "      <td>169631245</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                chr  startSite    endSite strand    TssSite\n",
       "geneName                                                   \n",
       "ENSG00000000003   X   99883667   99894988      -   99894988\n",
       "ENSG00000000005   X   99839799   99854882      +   99839799\n",
       "ENSG00000000419  20   49551404   49575092      -   49575092\n",
       "ENSG00000000457   1  169821804  169863408      -  169863408\n",
       "ENSG00000000460   1  169631245  169823221      +  169631245"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "groupbyTss.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "for row in groupbyTss.itertuples():\n",
    "    if row[2]>row[3]:\n",
    "        print(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'groupbyTss' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-c06bcf293a43>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mgroupbyTss\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mto_csv\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/home/shuang/projects/development_eqtm/data/geneSites/all_geneSites.txt'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'groupbyTss' is not defined"
     ]
    }
   ],
   "source": [
    "groupbyTss.to_csv('/home/shuang/projects/development_eqtm/data/geneSites/all_geneSites.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
